Add cmake section to knowledge base.

cmeyer opened the pull request:
<!--
Thank you for pull request.
-->

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
Supersedes https://github.com/conda-forge/conda-forge.github.io/pull/1015

Basic CMake build notes. I decided to make it different from the existing section since the existing section is Windows-specific. May consider merging/expanding the two sections in the future.

isuruf commented on src/maintainer/knowledge_base.rst:
```diff
@@ -791,6 +793,52 @@ Following implies that ``python`` is a runtime dependency and a Python matrix fo
+
+.. code-block:: yaml
+
+    host:
```
```suggestion
    build:
```

cmeyer replied:
Ok, I can change this -- although as I understand it, `host` by itself should be preferable to a separate `build` and `host` section. Am I mistaken?

isuruf replied:
Yes, `cmake` in `host` is wrong. `host` is usually headers, libraries, host. `build` is build tools like the compiler, make, cmake, pkg-config.

jakirkham commented:
WDYT about moving this before "Particularities on Windows"?

isuruf commented on src/maintainer/knowledge_base.rst:
```diff
@@ -791,6 +793,52 @@ Following implies that ``python`` is a runtime dependency and a Python matrix fo
+
+CMake requires you to supply a configuration file, usually called ``CMakeLists.txt``.
+You can use the one from your source tar ball if it exists or include your own in the
+same directory as the conda-forge recipe file and copy it to the source directory.
```
I don't think we need to encourage writing CMake configuration files for packages. Instead we should recommend that packages should use CMake if the upstream build system is CMake or if the upstream doesn't have a build system.

cmeyer replied:
Again, I'm ok changing this, and I'll go ahead and do that. But... there are a few reasons this might be useful "as is" or with  a slightly modified comment. Sometimes there is already a package release with a tar ball that has no CMake support or CMake that is not compatible with conda-forge. In this case, it's useful to know that you can include the config file near the recipe and use it, particularly when debugging a recipe. (I've been thinking of writing a 'debugging recipes' section too btw).

jakirkham replied:
I'm torn. There are people who did write their own `CMakeLists.txt` to get things to build in conda-forge in the early days. However I agree with Isuru, this is scope creep and it seems hard to stay on top of. Maybe we can link out to some CMake docs somewhere?

cmeyer replied:
I guess one question would be "who is the audience?" If it's msbuild users switching to cmake, then I think it should stay; but it sounds like the msbuild problem might be resolved soon anyway. So then it's more likely that the audience will be people adding projects with existing cmake builds.

cmeyer replied:
I removed this section too.

jakirkham replied:
I think that is the case (even outside of the msbuild problem).

In any event, we probably will wind up solving the msbuild problem by trying to migrate things to CMake if they already support CMake as opposed to adding that support for projects. Cases without CMake as an option will likely get solved in a different way (likely moving to vs2017).

cmeyer commented:
That would put it at the top of the document...? I'm ok with that if you are.

isuruf commented on src/maintainer/knowledge_base.rst:
```diff
@@ -791,6 +793,52 @@ Following implies that ``python`` is a runtime dependency and a Python matrix fo
+Other useful options are ``-B<directory>`` and ``-S<directory>`` to specify build and source
+directories.
+
+Inside the ``CMakeLists.txt`` file, the following options may be useful.
```
I don't think conda-forge documentation is the place for instructions on how to write a `CMakeLists.txt`

cmeyer replied:
Again, technically true, but there is enough searching around when figuring out how to use CMake that I would have found it profoundly useful to have.

cmeyer replied:
I removed this section.

jakirkham commented:
> That would put it at the top of the document...? I'm ok with that if you are.

SGTM. The Windows section seems to go into some CMake details. So it seem better to cover that first. Plus this has more general applicability (other OSes use CMake too).

jakirkham commented on src/maintainer/knowledge_base.rst:
```diff
@@ -791,6 +793,52 @@ Following implies that ``python`` is a runtime dependency and a Python matrix fo
+
+.. code-block::
+
+    cmake -G "NMake Makefiles" -DCMAKE_BUILD_TYPE:STRING=Release -DPython3_EXECUTABLE="%PYTHON%"
```
The `-G "NMake Makefiles"` is Windows specific. Should we put that bit in the Windows CMake section?

cmeyer replied:
It's already there. The reason I left it is that I didn't want to write a self contained example for Mac/Linux (just time constraints). And it seemed strange to have only the Mac/Linux example at that point. I'll think about that one.

cmeyer replied:
I left it in the latest revision, but added a note to refer to the other section too. Does this work for you?

jakirkham replied:
Yeah that seems fine. Thanks!

jakirkham commented:
Thought on using a `cmake_build` directory in the examples? Would say `build`, but Bazel likes to call its scripts `BUILD`, which has caused conflict before.

cmeyer commented:
I purposely didn't specify the build directory in these examples - adding it would be nice, but, again, would push towards more complete examples which I was hoping to avoid :)

mingwandroid commented:
Why do you propose adding it to the host env? This should not be done. Only ever add build tools to build please!

jakirkham commented:
It's in `build`.

mingwandroid commented:
Ah great. I'm blaming my lack of skills in the github app for this one ;-)

Apologies.

jakirkham commented:
No worries. That may have been an issue previously. I think it's been fixed now 🙂

xhochy commented:
Thanks for starting this, I would include the following things also:

### Flags that are always useful (I use these in 100% of all cases)

* `-DCMAKE_BUILD_TYPE=release`: Setup as a release build. This is better done on the initial `cmake` call as some packages build different build configurations depending on this flag. This not only adjust compile flags.
* `-DCMAKE_INSTALL_PREFIX=$PREFIX`: Install into the correct place.
* `-DCMAKE_INSTALL_LIBDIR=lib`: Libraries will land in `$PREFIX/lib`, sometimes projects install into `lib64` or similar but on conda-forge we keep shared libraries in simply `lib`.

### Helpful flags (I use these in 80% of all cases)

* `-DBUILD_TESTING=OFF`: Disable building the unit tests. We normally only test the existence of certain files and don't build the whole unit tests suite. Disabling it saves a lot of compute resources as well does it minimize the number of dependencies needed for the build.
* `-DBUILD_SHARED_LIBS=ON`: Instruct CMake to build shared libraries  instead of static ones.

### Other things

**Build tool:** Not only `cmake` needs to be specified in `build:` but also the tool to do the build, e.g. `make` or `ninja`. I tend to use `Ninja` as a `CMAKE_GENERATOR (-G)` nowdays exclusively as it utilises all CPUs automatically, works for Windows and Unix and gives a much better command line output for parallel builds than `make` does.

Happy to submit a follow-up PR or a PR to this PR for this.

ryanvolz commented:
Just a comment from an observer: the `-DPython3_EXECUTABLE=${PYTHON}` part is an absolutely necessary tip, but it might also be helpful to mention that this only works if the project is set up to use the new-ish [FindPython3 module](https://cmake.org/cmake/help/git-stage/module/FindPython3.html) through ``find_package(Python3)`` (and maybe just plain FindPython too through ``find_package(Python)``?). If the project uses the deprecated [FindPythonInterp module](https://cmake.org/cmake/help/latest/module/FindPythonInterp.html), this will not work. For that, it's necessary to pass `-DPYTHON_EXECUTABLE=${PYTHON}`. That's not to mention any projects that try to do custom detection for Python.

I'd just worry about it being more confusing if people read this, expect it to always work, and then can't figure out why it doesn't because the docs say it should.

cmeyer commented:
@xhochy Good suggestions. I added these to my documentation, although I didn't rank them as 100% vs 80% -- just that they might be useful.

@ryanvolz Another good suggestion. I added a qualifier about `FindPython3` and a note that older projects may need similar but different options.

jakirkham commented:
@xhochy, on the build tool point, what do you think about recommending [this]( https://github.com/conda-forge/symengine-feedstock/blob/ccf74a0da3948cf845591b3f2289241f64718cf5/recipe/build.sh#L21-L22
 )?

```
cmake --build . -- -j${CPU_COUNT}
cmake --build . --target install
```

xhochy commented:
> @xhochy, on the build tool point, what do you think about recommending [this](https://github.com/conda-forge/symengine-feedstock/blob/ccf74a0da3948cf845591b3f2289241f64718cf5/recipe/build.sh#L21-L22)?
> 
> ```
> cmake --build . -- -j${CPU_COUNT}
> cmake --build . --target install
> ```

Sounds reasonable as it works with all CMake Generators quite well.

cmeyer commented:
My recipe does not use those lines and they are not required for my recipe. So I'm not going to put "100%" on it. Feel free to commit what I have and submit a PR. And '--target install' specifically makes my recipe fail. It seems like bad guidance to say something is "100%".

cmeyer commented:
@isuruf Is there something else I should to do to get this approved/merged?

isuruf commented on src/maintainer/knowledge_base.rst:
```diff
@@ -1,6 +1,65 @@
+    - ``-DCMAKE_INSTALL_PREFIX=$PREFIX`` Specify the install location.
+    - ``-DCMAKE_INSTALL_LIBDIR=lib`` Libraries will land in $PREFIX/lib, sometimes projects install
+      into lib64 or similar but on conda-forge we keep shared libraries in simply lib.
+    - ``-DBUILD_TESTING=OFF`` Disable building the unit tests. We normally only test the existence of
```
Can you remove this recommendation? We encourage testing if possible within time constraints.

Otherwise looks good to me. Thanks for working on this.

cmeyer replied:
Done

xhochy commented:
@cmeyer out of curiosity, can you give me a link to your package?

cmeyer commented:
@xhochy One of my packages is [nionui-tool](https://github.com/nion-software/nionui-tool) (it is a high performance Qt application that dynamically embeds Python). Look in the `launcher` subdirectory for the `CMakeLists.txt` file. The [associated feedstock](https://github.com/conda-forge/nionui-tool-feedstock).

On Linux, I had a lot of difficulty getting cmake to produce a relocatable binary that would still find the Qt libraries. Eventually, I settled on letting cmake produce a non-relocatable binary and using the following custom post-build command (requires `chrpath` to be installed) to manually change the `rpath`:

```
    add_custom_command(
      TARGET ${PROJECT_NAME} POST_BUILD
      COMMAND chrpath -r $ORIGIN:$ORIGIN/../../lib
        $<TARGET_FILE_DIR:${PROJECT_NAME}>/NionUILauncher VERBATIM)
```

As an additional note: this application gets installed as a command line tool, but runs a full Qt application that dynamically loads the Python in which it is running (finding symbols using `dlsym`). It does not link to Python. Then it sets up the dynamically loaded Python environment (using `Py_SetPythonHome` and `Py_SetProgramName`). It will run within conda or venv virtual environments. It features low latency threaded drawing useful for live data acquisition (~10ms for small images; ~50ms for 2Kx2K images).
